### PR TITLE
Add `VocabMaskTransform`

### PR DESCRIPTION
**Pull request by shadeMe**
## Description

Add `VocabMaskTransform` logit transform that's unconditionally applied to all generations.

### Types of change

Enhancement

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: curated_transformers/generation/config.py [hunk 1]
```diff
@@ -21,29 +21,34 @@ class GeneratorConfig(ABC):
...
         when predicted. When this value is set to `None`, it is the
         responsibility of the generator to set it.
-
     :param max_generated_pieces:
         The maximum number of generation steps. This condition is a noop
         for values less than 1. When this value is set to `None`, it is the
         responsibility of the generator to set it.
     """
 
+    default_logits_transform: LogitsTransform = field(
```
Comment (danieldk) [inline curated_transformers/generation/config.py:38]: This is not how configuration has worked so far. In the current setup, we'd rather add a field like `masked_pieces: Set[int]` and the `logits_transform` method of the particular config would convert this to a `LogitsTransform`.

We should make it possible for the user to directly configure transforms (I have an idea of how we could do this, but this requires changing the class hierarchy, but that's out the scope of this PR). But the main idea behind the current configuration classes is that the user configures concrete options and does not have to bother with the underlying classes (like `LogitsTransform` subclasses).

File: curated_transformers/generation/logits.py [hunk 2]
```diff
@@ -121,3 +122,43 @@ def _process_logits(self, logits: Tensor):
             return
 
         logits /= self.temperature
+
+
+class MaskTransform(LogitsTransform):
```
Comment (danieldk) [inline curated_transformers/generation/logits.py:127]: I think it would be nice if the name also makes clear what we are masking, since we already have multiple masks. Maybe `MaskPiecesTransform`? `VocabMaskTransform`?

File: curated_transformers/generation/logits.py [hunk 3]
```diff
@@ -121,3 +122,43 @@ def _process_logits(self, logits: Tensor):
...
+    """
+    Set the probability of specific classes to zero.
```
Comment (danieldk) [inline curated_transformers/generation/logits.py:129]:
```suggestion
    Set the probability of specific vocabulary items to zero.
```

File: curated_transformers/generation/logits.py [hunk 4]
```diff
@@ -121,3 +122,43 @@ def _process_logits(self, logits: Tensor):
...
+    """
+
+    def __init__(self, classes_to_mask: List[int]):
```
Comment (danieldk) [inline curated_transformers/generation/logits.py:132]:
```suggestion
    def __init__(self, classes_to_mask: List[int]):
```

Probably `Set[int]` or if we want to be more generic `Iterable[int]`.

File: curated_transformers/generation/logits.py [hunk 5]
```diff
@@ -121,3 +122,43 @@ def _process_logits(self, logits: Tensor):
...
+        :param classes_to_mask:
+            Identifers pertaining to the classes that need to be masked.
+            An empty list results in a no-op.
+        """
+        super().__init__()
+        self.classes_to_mask = torch.tensor(
+            classes_to_mask, dtype=torch.long
+        )  # `torch.int32` isn't supported in indexing operations prior in torch<2.0.0.
+        if self.classes_to_mask.dim() != 1:
+            raise ValueError("Class selector tensor for mask transform must be 1D")
```
Comment (danieldk) [inline curated_transformers/generation/logits.py:145]: I think in the case of generation the classes are always vocab items, so I think it's clearer if the errors mention that. Something like:

```suggestion
            raise ValueError("Vocabulary items mask must be 1D")
```

the same of all the mentions of _class_ below.

File: curated_transformers/generation/logits.py [hunk 6]
```diff
@@ -121,3 +122,43 @@ def _process_logits(self, logits: Tensor):
...
+    def _process_logits(self, logits: Tensor):
+        if self.classes_to_mask.size(dim=-1) == 0:
+            return
+
+        max_expected = logits.size(dim=-1)
+        max_received = int(self.classes_to_mask.max(dim=-1).values)
+        if max_received >= max_expected:
+            raise ValueError(
+                f"Class identifiers for mask transform must be < {max_expected}, but got {max_received}"
+            )
```
Comment (danieldk) [inline curated_transformers/generation/logits.py:161]: I think Torch reports out of bound indexing. Good error messages are nice, but this does launch another kernel. I think it's worth looking wether the indexing below can be wrapped in a `try...except`, so that we don't have to go over the array twice.

File: curated_transformers/generation/config.py (hunk 1 above)
Comment (shadeMe) [reply curated_transformers/generation/config.py:38]: Makes sense. I'll rewrite it as suggested.

File: curated_transformers/generation/logits.py (hunk 2 above)
Comment (shadeMe) [reply curated_transformers/generation/logits.py]: Renamed to `VocabMaskTransform`.

File: curated_transformers/generation/logits.py (hunk 4 above)
Comment (shadeMe) [reply curated_transformers/generation/logits.py]: Switched to `Iterable`

File: curated_transformers/generation/logits.py (hunk 5 above)
Comment (shadeMe) [reply curated_transformers/generation/logits.py]: Fixed.

File: curated_transformers/generation/logits.py (hunk 6 above)
Comment (shadeMe) [reply curated_transformers/generation/logits.py]: Switched to `try...except` .

**Review by danieldk — APPROVED** [review]
